[ZEPPELIN-3653] - New Java interpreter

selvinsource opened the pull request:
### What is this PR for?
**New Java interpreter**
There are several Java libraries that could be used to leverage the Data Visualization & Collaboration features of Zeppelin hence the need of a Java interpreter to run any java code / library with no further dependencies.
Dependencies to any java library can be added by the end users in the Java Zeppelin interpreter Dependencies settings section.

### What type of PR is it?
* Feature

### Todos
* Any feedback from reviewers

### What is the Jira issue?
* [ZEPPELIN-3653]

### How should this be tested?
Manually
* Start the Zeppelin server
* Create a new note with the java interpreter binding
* Write some java code as per documentation (docs/interpreter/java.md)
Unit tests
* Run unit tests (JavaInterpreterTest.java and JavaInterpreterUtilsTest.java)

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update?
No, the dependency on com.thoughtworks.qdox was already added as part of the Beam Interpreter.
* Is there breaking changes for older versions?
No.
* Does this needs documentation?
Yes, it has been added to the PR, see docs/interpreter/java.md.

zjffdu commented on conf/interpreter-list:
```diff
@@ -29,6 +29,7 @@ hbase           org.apache.zeppelin:zeppelin-hbase:0.7.0                Hbase in
 ignite          org.apache.zeppelin:zeppelin-ignite_2.11:0.7.0          Ignite interpreter built with Scala 2.11
 jdbc            org.apache.zeppelin:zeppelin-jdbc:0.7.0                 Jdbc interpreter
 kylin           org.apache.zeppelin:zeppelin-kylin:0.7.0                Kylin interpreter
+java            org.apache.zeppelin:zeppelin-java:0.7.0                 Java interpreter
```
The version here seems outdated, I created https://issues.apache.org/jira/browse/ZEPPELIN-3660 for it

zjffdu commented on conf/zeppelin-site.xml.template:
```diff
@@ -331,13 +331,13 @@
 <property>
   <name>zeppelin.interpreter.group.order</name>
-  <value>spark,md,angular,sh,livy,alluxio,file,psql,flink,python,ignite,lens,cassandra,geode,kylin,elasticsearch,scalding,jdbc,hbase,bigquery,beam,groovy</value>
+  <value>spark,md,angular,sh,livy,alluxio,file,psql,flink,python,ignite,lens,cassandra,geode,kylin,elasticsearch,scalding,jdbc,hbase,bigquery,beam,groovy,java</value>
```
I also concern whether we should keep this property. I believe zeppelin.interpreter.default should be sufficient. We only care about the default interpreter.

zjffdu commented:
Thanks @selvinsource for the contribution. Have you thought about other implementation of java interpreter. like https://github.com/albertlatacz/java-repl and JShell of Java 9

selvinsource commented:
@zjffdu thanks for reviewing the PR.

Let me give you a bit of background before I answer your question on using another java interpreter.

First of all, I found lately Zeppelin an amazing tool to work with and in particular using the Spark/scala interpreter.
However I also started evaluating / using Hazelcast Jet, which is a similar streaming/batch engine, entirely written in Java, and I immediately had a need to have that in Zeppelin to do some visualization of the results.

I therefore started to look into how to create an interpreter for Jet and went through all the existing interpreter to get some ideas. I then realized that the Beam interpreter is just a java interpreter apart from the many dependencies on Beam/Google libraries.
A solution could have been to clone the Beam interpreter to create a Jet one with dependencies only on Hazelcast Jet but that didn't seem the best solution (when you start cloning code).

Hence my idea to create a Java interpreter (using what has been done and tested already with Beam) and have Beam (and eventually Hazelcast Jet) interpreters to simply extends the Java one with engine specific dependencies.

So, this PR is a start, to have a common Java interpreter that other Java API/engine interpreters can extend.

I agree the current implementation is not ideal as a whole class with a main method is needed to run some simple java code, the better approach is to use a Java REPL API.

The obvious choice is Java 9 JShell as it is the only one that is going to be supported in the long run and no need of other dependencies as it is shipped as part of the JDK. I don't think it's wise to use any other Java REPL as they won't compete with JShell which will be the standard that every IDE will integrate with.

However, the question back to you, is Zeppelin ready to be compiled with Java 9 and requiring Java 9 run time? Is it wise at this stage?
Talking from experience, in production environments I have seen rarely Java 9, the Java community is waiting for the next LTS release which is Java 11 (due in Sep 2018).

I would say to start with a Java interpreter like the one I proposed (re-using what's good was done for Beam), have Beam and eventually Jet extends the Java interpreter. Once Zeppelin will run on Java 11, the Java interpreter can be easily upgraded to use the JShell API therefore all other java based interpreters will have the same benefits (Beam, Jet, etc...).

Definitely the Java interpreter will be much more usable and immediate with a REPL, it just need to wait for the JShell to be available.

If you want I can already create a JIRA ticket to eventually update the Java interpreter to JShell as soon as this is available and I am happy to contribute to it: having seen the JShell API is going to be even simpler than the current REPL implementation that has been used in the Beam interpreter (but let's not forget that it works just fine!).

zjffdu commented:
Thanks @selvinsource Very glad that you willing to improve the current java interpreter when JShell is available in Java 11. It make sense to use the current java repl that Beam use, and improve it later. 
BTW, do you plan to create Hazelcast Jet interpreter later after this PR is completed ?

selvinsource commented:
@zjffdu as soon as this PR is merged in the master I will update the upstream master and start working on Jet interpreter, I have created a Jira ticket already (with a dependency on the Java interpreter).

Hazelcast Jet interpreter will be similar to the Beam one in this PR, it will simply implement JavaInterpreter and it will have the dependencies required (hazelcast-jet jar). What's more I am planning to add some more utils for Hazelcast for iMap/iCache/DAG/etc, see the JIRA description https://issues.apache.org/jira/browse/ZEPPELIN-3654.

When is the next zeppelin release due (0.9)? I can try to put both in it before that.

zjffdu commented:
There's no exact plan for 0.9.0, my rough plan is to release 0.9.0 in the next half year of 2018.

zjffdu commented on pom.xml:
```diff
@@ -821,6 +821,13 @@
       </modules>
     </profile>
 
+    <profile>
```
Could you just add java module directly instead of using profile ? We plan to remove all these interpreter specific profiles.

zjffdu commented on java/src/main/java/org/apache/zeppelin/java/JavaInterpreter.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class JavaInterpreter extends Interpreter {
+
+  Logger logger = LoggerFactory.getLogger(JavaInterpreter.class);
```
Use final static

zjffdu commented on java/README.md:
```diff
@@ -0,0 +1,24 @@
+Current interpreter implementation supports the static repl. It compiles the code in memory, execute it and redirect the output to zeppelin.
+
+## Building the Java Interpreter
+You have to first build the Java interpreter by enable the **java** profile as follows:
```
Update this after you removed java profile

selvinsource commented:
@zjffdu 
I made the requested changes except zeppelin.interpreter.group.order, should I just remove the whole property? I can't see where it's used to be honest.

zjffdu commented on java/src/main/java/org/apache/zeppelin/java/JavaInterpreterUtils.java:
```diff
@@ -0,0 +1,44 @@
+   * @param rows Map of keys and values
+   * @return Zeppelin %table
+   */
+  public static String displayTableFromSimpleMap(String keyName, String valueName, Map<?, ?> rows){
```
It is fine to use JavaInterpreterUtils for now. But I think it is better to support ZeppelinContext in JavaInterpreter as well. So that user can use z.show to leverage the builtin visualization of zeppelin and also other features provided by ZeppelinContext.

zjffdu replied:
Sound good plan. Thanks

zjffdu commented:
@selvinsource I have removed property `zeppelin.interpreters`, could you do rebase ?

selvinsource commented:
@zjffdu I have done the rebase.

zjffdu commented on java/src/main/java/org/apache/zeppelin/java/JavaInterpreter.java:
```diff
@@ -0,0 +1,99 @@
+    File dir = new File(".");
+    // delete all .class files created while compilation process
+    for (int i = 0; i < dir.list().length; i++) {
+      File f = dir.listFiles()[i];
```
move dir.listFiles out of this for loop, otherwise listFiles will be called multiple times.

selvinsource replied:
OK, good point, I will change it shortly.

felixcheung commented on java/pom.xml:
```diff
@@ -0,0 +1,88 @@
+    <dependency>
+      <groupId>com.thoughtworks.qdox</groupId>
+      <artifactId>qdox</artifactId>
+      <version>2.0-M3</version>
```
LICENSE update?

felixcheung replied:
ok I saw the PR comment on this

selvinsource replied:
Yes, it was already updated as part of the Beam interpreter.

felixcheung commented on docs/interpreter/java.md:
```diff
@@ -0,0 +1,54 @@
+<div id="toc"></div>
+
+## How to use
+Basically, you can write normal java code. You should write the main method inside a class because the interpreter invoke this main to execute the code. Unlike Zeppelin normal pattern, each paragraph is considered as a separate job, there isn't any relation to any other paragraph.
```
perhaps clarify " each paragraph is considered as a separate job, there isn't any relation to any other paragraph."
does it mean I can't share variables in different paragraphs? I think it will greatly help if this is documented

selvinsource replied:
At the moment there isn't a way to share variable unless you use a clustered environment.
For example, when I will implement the Hazelcast interpreter, one could store a variable in the hazelcast cluster as part of one job and retrieve the same variable as part of another job.

As in previous comments with @zjffdu when we will move to JShell it will be easier to introduce the support for ZeppelinContext which will allow also to share variables.

felixcheung commented on java/README.md:
```diff
@@ -0,0 +1,17 @@
+Java interpreter for Apache Zeppelin
+
+# Architecture
+Current interpreter implementation supports the static repl. It compiles the code in memory, execute it and redirect the output to zeppelin.
```
`repl` -> `REPL`

selvinsource replied:
Ok I will update this.

selvinsource commented:
@zjffdu @felixcheung I have pushed the requested changes.

zjffdu commented:
LGTM, will merge soon

felixcheung commented:
did you add the behavior in doc?

selvinsource commented:
@zjffdu thanks

@felixcheung should I mention explicitly that one cannot share variables between paragraph? I think other interpreters do not allow that either but it's not mentioned. Instead, in the interpreter that is supported (like Spark) in the doc it mentions one could use the ZeppelinContext.
Also can you confirm that when you say "share variable" you are talking about the ZeppelinContext and its dynamic form support?
Let me know, I can explicitly add that if you prefer.

felixcheung commented:
no, I mean

paragraph 1
```
%java

Map<String, Long> counts = new HashMap<>();
```

paragraph 2
```
%java

counts.put("key", 1);
```

selvinsource commented:
@felixcheung 
I see what you mean now and I added explicitly to the Java Interpreter doc that the Java interpreter do not support it.
Eventually, when we move the Java interpreter implementation to JShell we can look into adding a JavaContext that allows to share variables between paragraph as you mentioned.

@zjffdu 
I have rebased once more for an easy merge to the master.

zjffdu commented:
@selvinsource What @felixcheung mentioned is not sharing varaibles via JavaContext, he mean share variable in the same java repl as spark repl did right now.

selvinsource commented:
@zjffdu yeah I think I understood and JShell should be able to achieve this.
I meant as JavaContext a way to maintain the same JShell instance to run multiple snipper across paragraph in a similar way (conceptually) the Spark interpreter does by using the SparkContext.
Definitely the aim when introducing JSchell is to support both sharing variables as in the example from @felixcheung as well as using the ZeppelinContext for Zeppelin input/output support. At that point the Java interpreter will have comparable features to the other interpreters. As I said I will add a Jira for this. My next step is to add the HazelCast Jet interpreter once this PR is merged.

zjffdu commented:
@selvinsource Travis still fails, I will commit after travis is passed

```

[ERROR] Failed to execute goal on project zeppelin-beam: Could not resolve dependencies for project org.apache.zeppelin:zeppelin-beam:jar:0.9.0-SNAPSHOT: Could not find artifact org.apache.zeppelin:zeppelin-java:jar:0.9.0-SNAPSHOT in sonatype-snapshots (https://oss.sonatype.org/content/repositories/snapshots/) -> [Help 1]
```

felixcheung commented:
right. just want to clarify that it is perfectly fine to run Scala, Python etc without Spark. The session context is maintained in the REPL implementation and not in SparkContext or ZeppelinContext. I think they are orthogonal.

zjffdu commented:
@selvinsource I think you need to update this line of .travis.xml  https://github.com/apache/zeppelin/blob/master/.travis.yml#L44

selvinsource commented:
@zjffdu I have made the change, if you can re-run Jenkins please.

zjffdu commented:
@selvinsource Do you set up travis ? https://zeppelin.apache.org/contribution/contributions.html

selvinsource commented:
@zjffdu yes I did before opening the pull request.

I manually re-run it and it's fine now
https://travis-ci.org/selvinsource/zeppelin/builds/411153675

zjffdu commented:
merging

selvinsource commented:
@zjffdu @felixcheung thanks for reviewing this PR and merging it in the master. I am going to work on the Hazelcast Jet interpreter next and send a new PR once ready.

mebelousov commented:
Seems this PR broke build for scala 2.11

After execution of ` ./dev/change_scala_version.sh 2.11` I have got error 
```
[ERROR] Failed to execute goal on project zeppelin-beam: Could not resolve dependencies for project org.apache.zeppelin:zeppelin-beam:jar:0.9.0-SNAPSHOT: Could not find artifact org.apache.zeppelin:zeppelin-scio_2.10:jar:0.9.0-SNAPSHOT in apache.snapshots (http://repository.apache.org/snapshots) -> [Help 1]```

@selvinsource could check maven build?

selvinsource commented:
@mebelousov 

If you see
https://travis-ci.org/selvinsource/zeppelin/builds/411153675
it seems fine with SCALA_VER="2.11".

What options are you running maven build with after the change of scala version? I could try to re-create in dev environment.

selvinsource commented:
@zjffdu 
the issue reported by @mebelousov may have been caused by https://issues.apache.org/jira/browse/ZEPPELIN-3557
and not this PR, as I only added in the beam interpreter the dependency on the java interpreter.

If you look at scio/pom.xml it has
```
  <groupId>org.apache.zeppelin</groupId>
  <artifactId>zeppelin-scio_2.10</artifactId>
  <packaging>jar</packaging>
  <version>0.9.0-SNAPSHOT</version>
  <name>Zeppelin: Scio</name>
  <description>Zeppelin Scio support</description>
```
So maven will always only build the 2.10 version for scio.

At the same time the beam/pom.xml
has dependency to
```
 <dependency>
      <groupId>org.apache.zeppelin</groupId>
      <artifactId>zeppelin-scio_${scala.binary.version}</artifactId>
      <version>${project.version}</version>
    </dependency>
```
which do not exists for scala 2.11.

So the solution is either that scio is also built for 2.11 or beam to reference 2.10 no matter what the scala version is.

I think you never noticed it before because the beam interpreter was "opt in" so perhaps nobody noticed this as they only ever built the beam interpreter (-Pbeam) with scala 2.10 and not 2.11.

But now that all profiles have been removed as per ZEPPELIN-3557, the beam interpreter needs the correct scio version.

selvinsource commented:
@zjffdu @mebelousov I sent you a comment before, please ignore, I thought that scio was only built for 2.10 but I noticed that changing scala version using the script does update the pom for scio to 2.11.

I still think is somehow related to https://issues.apache.org/jira/browse/ZEPPELIN-3557 as before people wouldn't build using profiles by default so there may be some issue uncovered now that all interpreters are built all the times.

In this PR I only changed beam/pom.xml to have a dependency on java interpreter, nothing was changed related to scio or scala versions.

Will investigate further but it would be good to know which options @mebelousov is using to build to re-create locally.

selvinsource commented:
Found something that could cause the issue.
in beam/pom.xml there is
```<beam.beam.version>2.0.0</beam.beam.version>```

then 
     ``` <artifactId>beam-runners-flink_${scala.binary.version}</artifactId>```

But looking at
https://mvnrepository.com/artifact/org.apache.beam/beam-runners-flink
there isn't a beam-runners-flink_2_10 for beam version 2.0.0.

So even in Zeppelin 0.8.0 if you try to switch to scala 2.11 and run install -DskipTests -Pbeam you will get an error but perhaps that's something nobody ever tried (to compile beam on 2.11).

However after removing all profiles, that's the case now and a solution is needed.

Either the beam version is updated or the beam-runners flink can be 2_10 even for scala 2.11.


@zjffdu what do you think about my analysis? Am I making sense?

zjffdu commented:
Thanks @selvinsource for the investigation, the analysis seems correct to me.

felixcheung commented:
so we need to open a JIRA then?

selvinsource commented:
@felixcheung yes I think so as it's not something new, it has been there since the beam interpreter was created but nobody noticed until now that all interpreter specific profiles have been removed.

@mebelousov for now the workaround is simple to change beam/pom.xml to
```<artifactId>beam-runners-flink_2.10</artifactId>```
before you build zeppelin.

selvinsource commented:
@felixcheung @zjffdu 
FYI, I have added https://issues.apache.org/jira/browse/ZEPPELIN-3742 based on the discussion we had in this PR.

selvinsource commented:
@zjffdu @felixcheung do you have any estimate on when version 0.9 is going to be released? Looking forwawrd to see the Java interpreter included in the downloadable version.

felixcheung commented:
oh wait, I didn't realize this comment https://github.com/apache/zeppelin/pull/3092#issuecomment-410957256 is not related to ZEPPELIN-3742 

do we need a fix for `beam-runners-flink_2.10`

selvinsource commented:
> oh wait, I didn't realize this comment [#3092 (comment)](https://github.com/apache/zeppelin/pull/3092#issuecomment-410957256) is not related to ZEPPELIN-3742
> 
> do we need a fix for `beam-runners-flink_2.10`

As I mentioned before, that's nothing new, it's been there since the Beam interpreter was added, it never supported scala 2.11, see the notes on
https://github.com/apache/zeppelin/blob/master/beam/README.md
"Flink runner comes with binary compiled for scala 2.10. So, currently we support only Scala 2.10".

By doing a quick search somebody else already created JIRAs for this
https://issues.apache.org/jira/browse/ZEPPELIN-3903
https://issues.apache.org/jira/browse/ZEPPELIN-3899

@novelari perhaps is in the best position to update the beam-runners-flink to a most up to date version to support both scala 2.10 and 2.11, I am not that familiar with Beam/Flink.

A solution could be to use profiles in beam/pom.xml to activate either 
https://mvnrepository.com/artifact/org.apache.beam/beam-runners-flink_2.10/2.2.0
or
https://mvnrepository.com/artifact/org.apache.beam/beam-runners-flink_2.11/2.3.0
depending on the scala version.

For example
<code>
<profiles>
    <profile>
      <id>scala-2.10</id>
      <activation>
        <activeByDefault>true</activeByDefault>
      </activation>
        <dependencies>
<dependency>
    <groupId>org.apache.beam</groupId>
    <artifactId>beam-runners-flink_2.10</artifactId>
    <version>2.2.0</version>
</dependency>
        </dependencies>
    </profile>
    <profile>
      <id>scala-2.11</id>
        <dependencies>
<dependency>
    <groupId>org.apache.beam</groupId>
    <artifactId>beam-runners-flink_2.11</artifactId>
    <version>2.3.0</version>
</dependency>
        </dependencies>
    </profile>
</profiles>
</code>

felixcheung commented:
ah yes, thanks for looking into this. it's a long standing problem with scala version in scala builds. I think it's a good idea to upgrade beam

sgloutnikov commented:
@selvinsource thank. you! Saved me a good amount of time investigating why I couldn't build with Scala 2.11. Was able to build fine with `beam-runners-flink_2.11:2.3.0`.
